### ISSUE TITLE
PUT keytab 500 error

### ISSUE THREAD

**Issue by dddugan** [hashicorp/vault-plugin-auth-kerberos#36]
Seem to be having the same problem as https://github.com/hashicorp/vault-plugin-auth-kerberos/issues/15, which was closed for inactivity.

Would appreciate any guidance on how to troubleshooting this.

```
$ VAULT_NAMESPACE="sandbox" vault write auth/kerberos/config keytab=@vault.keytab.base64 service_account="vault"
Error writing data to auth/kerberos/config: Error making API request.

URL: PUT https://myservername.tld:8200/v1/auth/kerberos/config
Code: 500. Errors:

* 1 error occurred:
        * internal error

$
```

**Comment by tyrannosaurus-becks** [comment]
Hi @dddugan , happy to take a look.

This plugin hasn't been released as part of Vault yet, though it will go out in the 1.4 release. Its docs are located [here](https://github.com/hashicorp/vault/pull/7993/files) and after the release will be much more user-friendly. 

Particularly, [this doc](https://github.com/hashicorp/vault/pull/7993/files#diff-8b8c7800989f1410f4b3be75dfc65d32R10) gives a walk-through of setting it up, and at the bottom there's a Troubleshooting section. The troubleshooting consists of checking the LDAP config using `ldapsearch`, and the KDC config by logging in with a keytab _outside_ of Vault.

Can you take a look at that, and post the results you find from troubleshooting? Also, are you on a version of the plugin build from the current `master` branch of this repo?

Thanks!

**Comment by dddugan** [comment]
Thanks, @tyrannosaurus-becks. I get through the setup just fine through registering the plugin in the catalog, enabling the auth method, and (outside Vault) creating the keytab. But I can never write auth/kerberos/config to upload the keytab and service_account information to Vault. So this seems like more of a plugin execution/environment issue than Kerberos/LDAP, but I'm happy to be shown otherwise, or if there's a better place to post please direct me.

That said, outside of Vault I can successfully get a ticket and query ldap like so:

```
$ kdestroy
$ kinit -k -t vault.keytab vault@MYDOMAIN.TLD
$ klist
Ticket cache: KEYRING:persistent:13579:krb_ccache_72yPKbF
Default principal: vault@MYDOMAIN.TLD

Valid starting       Expires              Service principal
01/30/2020 12:56:00  01/30/2020 22:56:00  krbtgt/MYDOMAIN.TLD@MYDOMAIN.TLD
        renew until 02/06/2020 12:56:00
$
$ ldapsearch -Y GSSAPI -h dc01.MYDOMAIN.TLD -p 389 -b "ou=users,dc=mydomain,dc=tld" samaccountname="user1" userPrincipalName -LLL
SASL/GSSAPI authentication started
SASL username: vault@MYDOMAIN.TLD
SASL SSF: 256
SASL data security layer installed.
dn: CN=user1,OU=Users,DC=mydomain,DC=tld
userPrincipalName: user1@mydomain.tld

$
```

I'll look at upping the Vault log level to see what else I can come up with. Does this affect plugin logging?

**Comment by dddugan** [comment]
May have found the issue related to cert name/IP mismatch. On startup we (always) see this:
```
Jan 30 13:53:42 myserver systemd: Started "HashiCorp Vault - A tool for managing secrets".
Jan 30 13:53:43 myserver vault: ==> Vault server configuration:
Jan 30 13:53:43 myserver vault: Api Address: https://12.34.56.78:8200
Jan 30 13:53:43 myserver vault: Cgo: disabled
Jan 30 13:53:43 myserver vault: Cluster Address: https://12.34.56.78:8201
Jan 30 13:53:43 myserver vault: Listener 1: tcp (addr: "0.0.0.0:8200", cluster address: "0.0.0.0:8201", max_request_duration: "1m30s", max_request_size: "33554432", tls: "enabled")
Jan 30 13:53:43 myserver vault: Log Level: trace
Jan 30 13:53:43 myserver vault: Mlock: supported: true, enabled: true
Jan 30 13:53:43 myserver vault: Storage: consul (HA available)
Jan 30 13:53:43 myserver vault: Version: Vault v1.1.3+prem
Jan 30 13:53:43 myserver vault: Version Sha: af0b27edfd5bcd9123b6ac69e90c6e6dfa3654f7
Jan 30 13:53:43 myserver vault: ==> Vault server started! Log data will stream in below:
Jan 30 13:53:43 myserver vault: 2020-01-30T13:53:43.025-0600 [WARN]  no `api_addr` value specified in config or in VAULT_API_ADDR; falling back to detection if possible, but this value should be manually set
```

And after enabling trace logging and trying to write auth/kerberos/config I see the following that looks relevant. Indeed our cert does not include IPs - only names. Should api_addr be set on each vault node with its own fqdn, or the fqdn of the load balancer? Something else?
```
Jan 30 14:13:14 myserver vault: 2020-01-30T14:13:14.311-0600 [ERROR] auth.kerberos.auth_kerberos_dfe36b16.kerberos.vault-plugin-auth-kerberos: plugin tls init: error="error during token unwrap request: Put https://12.34.56.78:8200/v1/sys/wrapping/unwrap: x509: cannot validate certificate for 12.34.56.78 because it doesn't contain any IP SANs" timestamp=2020-01-30T14:13:14.310-0600
```

**Comment by MrFixit96** [comment]
Following https://www.vaultproject.io/docs/concepts/ha/#direct-access as long as clients can access the individual nodes, set the api_addr on each node to its https://fqdn:8200 or https://x.x.x.x:8200 and if need be the node will forward any writes to the active node.

tldr; if you want to use IP's add the appropriate IP SANs to your cert... else, set your urls with dns names in each node.

**Comment by dddugan** [comment]
Thanks, @MrFixit96. Setting api_addr appropriately on each node indeed solved the issue. (setting to https://f.q.d.n:8200) Continuing configuration and testing, but not a fundamental plugin problem. Closing the issue.

**Comment by tyrannosaurus-becks** [comment]
@dddugan Thanks for posting about all this! I did have one last question, what version of the plugin are you using? Is it from master of this repository? If so I'd like to make the error message returned more useful for debugging, that's why I'm asking.

**Comment by dddugan** [comment]
Ho hum...was actually using the wintoncode binary, not the Hashi stuff. Thought I'd go there if needed, but was trying for a drop-in until Hashi releases and we update to 1.4. In hindsight I probably should have started on the Google Group forum rather than github, but appreciate your prompt responses!
